### ISSUE TITLE
DNS leak during DHT bootstrap over SOCKS

### ISSUE THREAD

**Issue by ichorid** [arvidn/libtorrent#6295]
Libtorrent 1.2.7 in Tribler on Windows and Linux. Libtorrent is using SOCKS5 proxy in Tribler.
The same thing happens with `client_test`.

![изображение](https://user-images.githubusercontent.com/2509103/123433791-1ce8de00-d5cc-11eb-8f2e-6cf80172082c.png)

The rest of the traffic goes through the tunnel normally.

**Comment by arvidn** [comment]
this is not as simple as it might seem. I can't think of a way to actually resolve a name via SOCKS5, all you can do is to specify a hostname instead of an IP address when sending packets. The DHT routing table contains IP addresses, not hostnames. Currently there's an extra step to resolve names into IPs before adding them to the routing table.

I'm hesitant to make the routing table able to hold hostnames.

**Comment by ichorid** [comment]
There is a setting called `dht_bootstrap_nodes`. Documentation states:
>This is a comma-separated list of IP port-pairs. They will be added to the DHT node (if it's enabled) as **back-up nodes** in case we don't know of any.

Is there a way to completely override the addresses of Mainline DHT nodes? At least in Tribler, we could start providing bootstrap nodes through our own means.

**Comment by arvidn** [comment]
yes, if you put IP addresses in that setting, there won't be any hostname lookups.

**Comment by arvidn** [comment]
I started exploring deferring hostname lookups to the DHT socket class, which supports SOCKS5 hostnames in this branch `dht-hostname-lookups`. It's far from complete and I think I should simplify the DHT implementation a bit first, to avoid the complexity from growing out of control

**Comment by arvidn** [comment]
> How can one verify that DHT queries are correctly sent via socks?

Wireshark is probably the best tool

**Comment by arvidn** [comment]
all (public) torrents are announced to the DHT.
